pandas/core/tools/datetimes.py

```diff
@@ -557,7 +559,48 @@ def _to_datetime_with_unit(arg, unit, name, utc: bool, errors: str) -> Index:
         tz_parsed = None
     else:
         arg = np.asarray(arg)
-        arr, tz_parsed = tslib.array_with_unit_to_datetime(arg, unit, errors=errors)
+
+        if arg.dtype.kind in ["i", "u"]:
+            # Note we can't do "f" here because that could induce unwanted
+            #  rounding GH#14156, GH#20445
+            arr = arg.astype(f"datetime64[{unit}]", copy=False)
+            try:
+                arr = astype_overflowsafe(arr, np.dtype("M8[ns]"), copy=False)
+            except OutOfBoundsDatetime:
+                if errors == "raise":
+                    raise
+                arg = arg.astype(object)
+                return _to_datetime_with_unit(arg, unit, name, utc, errors)
+            tz_parsed = None
+
+        elif arg.dtype.kind == "f":
+            mult, _ = precision_from_unit(unit)
+
+            iresult = arg.astype("i8")
+            mask = np.isnan(arg) | (arg == iNaT)
```

Comment:
> are we confident that np.isnan(arg) addresses the issue for which values != values needed to be added?

im pretty confident, yes

> why could it not have been used in tslib.pyx? I'm guessing it was for the performance penalty of calling a Python function within Cython?

it should work just fine in cython too.

Reply:
OK let's leave that to another PR, easier to revert in case something goes wrong

---

also, should this be
```suggestion
            mask = np.isnan(arg) | (iresult == iNaT)
```
?

Reply:
the only test which seems to require that is `pandas/tests/tools/test_to_datetime.py::TestToDatetimeUnit::test_to_datetime_unit_fractional_seconds`, and it doesn't seem to make a difference there so maybe it's OK